tests/integration/test_healing.py

```diff
@@ -0,0 +1,298 @@
+    wait_network_restore,
+)
+from pytest_operator.plugin import OpsTest
+from tenacity import RetryError, Retrying, stop_after_attempt, wait_fixed
+
+from src.constants import CLUSTER_ADMIN_USERNAME, SERVER_CONFIG_USERNAME
+
+logger = logging.getLogger(__name__)
+
+METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
+APP_NAME = METADATA["name"]
+MYSQL_DAEMON = "mysqld"
+
+
+async def build_and_deploy(ops_test: OpsTest) -> None:
+    """Build and deploy."""
+    if app := await app_name(ops_test):
+        if len(ops_test.model.applications[app].units) == 3:
+            return
+        else:
+            async with ops_test.fast_forward():
+                await scale_application(ops_test, app, 3)
+            return
+
+    # Build and deploy charm from local source folder
+    charm = await ops_test.build_charm(".")
+
+    # Reduce the update_status frequency until the cluster is deployed
+    async with ops_test.fast_forward():
+        await ops_test.model.deploy(charm, application_name=APP_NAME, num_units=3)
+
+        await ops_test.model.block_until(
+            lambda: len(ops_test.model.applications[APP_NAME].units) == 3
+        )
+        await ops_test.model.wait_for_idle(
+            apps=[APP_NAME],
+            status="active",
+            raise_on_blocked=True,
+            timeout=1000,
+        )
+
+
+@pytest.mark.order(1)
+@pytest.mark.abort_on_fail
+@pytest.mark.healing_tests
+async def test_kill_db_process(ops_test: OpsTest) -> None:
+    """Kill mysqld process and check for auto cluster recovery."""
+    await build_and_deploy(ops_test)
+
+    app = await app_name(ops_test)
+
+    primary_unit = await get_primary_unit_wrapper(ops_test, app)
+    another_unit = (set(ops_test.model.applications[app].units) - {primary_unit}).pop()
+
+    primary_unit_ip = await get_unit_ip(ops_test, primary_unit.name)
+
+    # get running mysqld PID
+    pid = await get_process_pid(ops_test, primary_unit.name, MYSQL_DAEMON)
+
+    # kill mysqld for the unit
+    logger.info(f"Killing process id {pid}")
+    await ops_test.juju("ssh", primary_unit.name, "sudo", "kill", "-9", pid)
+
+    config = {
+        "username": CLUSTER_ADMIN_USERNAME,
+        "password": await get_system_user_password(primary_unit, CLUSTER_ADMIN_USERNAME),
+        "host": primary_unit_ip,
+    }
+
+    # retrieve new PID
+    new_pid = await get_process_pid(ops_test, primary_unit.name, MYSQL_DAEMON)
+    logger.info(f"New process id is {new_pid}")
+
+    # verify that mysqld instance is not the killed one
+    assert new_pid != pid, "❌ PID for mysql daemon did not changed"
+
+    # verify daemon restarted via connection
+    assert is_connection_possible(config), f"❌ Daemon did not restart on unit {primary_unit.name}"
+
+    # verify instance is part of the cluster
+    logger.info("Check if instance back in cluster")
+    assert await is_unit_in_cluster(
+        ops_test, primary_unit.name, another_unit.name
+    ), " Unit not online in the cluster"
```

Comment:
You'll want to verify that while the original primary was down that verify a new primary got elected.

The check for "Write to the cluster while old primary is down, verify it is up to date when it comes back online." can be done when this is sorted ("There's some limitations on doing tests under load since a client will connect to one instance at the time (as opposed to mongo replica set). This can be changed in the future by using mysql-router charm (which is not fully ready) , which will take care of bounce the client connection between cluster instances.")